### ISSUE TITLE
Modules ignore `subscription_id` when using CLI auth; resources get created in the wrong subscription.

### ISSUE THREAD

**Issue by UnwashedMeme** [ansible/ansible_collections_azure#181]
See https://github.com/Azure/azure_preview_modules/issues/321 and #53. 

To me this is a basic WTF issue. If I can't specify where a resource gets created properly then how am I supposed to do any serious automation?  There has been a pull-request present for 4months #53. Please fix this already.

This is especially frustrating when dealing with things that span multiple subscriptions. E.g. I want to look up vnet facts in one subscription to use it for a VM on the other subscription. Or I have a playbook that is creating a VM on each subscription.  Or I want to assign a VM identity permission to a resource in a different subscription.

A playbook, e.g.
```
- name: Create Azure VMs
  strategy: free
  hosts:
    - localhost
  gather_facts: false

  tasks:
    - name: "Create Resource group"
      delegate_to: localhost
      azure.azcollection.azure_rm_resourcegroup:
        name: "foobar"
        location: "eastus2"
        auth_source: cli
        subscription_id: "ecb1bdd8-..."
```
Ignores the `subscription_id` field:

```
$ az account list --verbose --query '[*].{id:id,isDefault:isDefault,state:state}'
[
  {
    "id": "f1c043c7-...",
    "isDefault": false,
    "state": "Enabled"
  },
  {
    "id": "ecb1bdd8-...",
    "isDefault": true,
    "state": "Enabled"
  }
]

$ ansible-playbook -v subtest.yml

TASK [Create Resource group] **********************************************************************************************************************
task path: /workspaces/sqlserver-deployment/subtest.yml:7
changed: [localhost] => changed=true 
  contains_resources: false
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      append_tags: true
      auth_source: cli
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      force_delete_nonempty: false
      location: eastus2
      name: foobar
      password: null
      profile: null
      secret: null
      state: present
      subscription_id: f1c043c7-...
      tags: null
      tenant: null
  state:
    id: /subscriptions/ecb1bdd8-.../resourceGroups/foobar
    location: eastus2
    name: foobar
    provisioning_state: Succeeded
    tags: null```

Note that even though I specifically asked for the rg to be created in "f1c043c7-..." it was created in "ecb1bdd8-..." .

My work around at the moment is: 
```
    - name: "Ensure the correct subscription"
      delegate_to: localhost
      shell: "az account set -s {{ subscription }}"
```
but this fails when trying to run for multiple inventory items simultaneously (e.g. build 1 vm on each vm).

**Comment by UnwashedMeme** [comment]
I didn't re-create this issue from the old repo to this repo immediately upon development being transitioned over because "There's already a pull-request, it will be fixed soon; I'll try to keep the noise down."

At this point it just came up, again, as a blocker. So I'm hear to make some noise about it again.   @Fred-sun, please get this fix  through.

**Comment by Fred-sun** [comment]
> I didn't re-create this issue from the old repo to this repo immediately upon development being transitioned over because "There's already a pull-request, it will be fixed soon; I'll try to keep the noise down."
> 
> At this point it just came up, again, as a blocker. So I'm hear to make some noise about it again. @Fred-sun, please get this fix through.

Thanks for your update! I will check it!

**Comment by UnwashedMeme** [comment]
For other people coming through, still waiting on this to be fixed, here's a workaround sketch:
```
  block:
  - name: "Switch to other subscription"
    delegate_to: localhost
    check_mode: false
    command: "az account set -s 'FOOBAR'"
  - name: "Grant access"
    delegate_to: localhost
    azure_rm_roleassignment: 
       subscription_id: "foobar"  # THIS STILL DOESN'T WORK
       ...
  - name: "Switch back to original subscription"
    delegate_to: localhost
    check_mode: false
    command: "az account set -s ORIGINAL"
```

**Comment by Fred-sun** [comment]
@UnwashedMeme Thank you for reporting this problem, it should not be a Bug, is the parameter auth_source design. If you do not specify auth_source, the specified subscription_ID will be used. You can refer to the following information, thank you!
```e.g.
auth_source:           # Controls the source of the credentials to use for authentication. If not specified, ANSIBLE_AZURE_AUTH_SOURCE environment
                               variable will be used and default to `auto' if variable is not defined. `auto' will follow the
                               default precedence of module parameters -> environment variables -> default profile in credential
                               file `~/.azure/credentials'. When set to `cli', the credentials will be sources from the default
                               Azure CLI profile. Can also be set via the `ANSIBLE_AZURE_AUTH_SOURCE' environment variable. When
                               set to `msi', the host machine must be an azure resource with an enabled MSI extension.
                               `subscription_id' or the environment variable `AZURE_SUBSCRIPTION_ID' can be used to identify the
                               subscription ID if the resource is granted access to more than one subscription, otherwise the
                               first subscription is chosen. The `msi' was added in Ansible 2.6.

**Comment by UnwashedMeme** [comment]
**The main issue is that I currently cannot specify the subscription**   I can update the title of this issue, or close it and make a new one titled that way. I was trying to be precise about that I expected "That subscription_id should direct the subscription id that is effected by this ansible module".  

* `az` command line lets you specify which subscription you want to use per command
* powershell lets you specify a profile for which subscription you want to use per command  (I think from briefly skimming the docs, i haven't used it much)
* terraform lets you specify a provider for which subscription you want to use per resource
* python lets you specify a `subscription_id` for the client
* _ansible doesn't currently let you specify a subscription_id for the module unless you are using a service principal._

I still think #53 looks pretty good (though perhaps it should update this bit of docs) and using the `subscription_id` parameter to specify which subscription this module makes a lot of sense to me (as is already reflected in the line of the docs that says "[it] can be used to identify the subscription ID if the resource is granted access to more than one subscription")  

In my testing #53  does solve this issue quite nicely.  If you think instead that switching between subscriptions shouldn't be done with the `profile` parameter we could do that instead, the MR would look a lot like #53. I feel like that would be much more confusing ("Why doesn't subscription_id work?")

We could create an entirely new parameter named, `subscription` or `subscription_to_use` or `target_subscription` or `${module_name}_subscription`  but that all adds cognitive overhead of trying to figure out how this is different than `subscription_id` which, again from the docs posted above, says "can be used to identify the subscription ID if the resource is granted access to more than one subscription."

**Comment by UnwashedMeme** [comment]
I said above: "...unless  you are using a service principal".  reading [the current docs](https://github.com/ansible-collections/azure/blob/8aad2a0c43d16cf0f5619164751e1e7049316f69/plugins/doc_fragments/azure.py):

* subscription_id with MSI makes sense -- same reason as it would for the CLI, you have permission to more than one, which one should you use?
* with service principal: "To authenticate via service principal, pass subscription_id, client_id, secret and tenant." This doesn't make sense directly: you don't authenticate service principals using a `subscription_id` in the azure api. However once authenticated the subscription_id [is used for creating the management client correctly](https://github.com/ansible-collections/azure/blob/d8ce59810f7fc991b273ffb8ac47e96aa58c3287/plugins/module_utils/azure_rm_common.py#L830)

both of these leave us back at the original question of "Why doesn't this work for cli auth?" [It could](#53)

**Comment by Fred-sun** [comment]
@UnwashedMeme Thank you for your reply. I will re-investigate this question.

**Comment by UnwashedMeme** [comment]
@Fred-sun  i spent the last couple hours digging into that module. I tried to improve the documentation and then I sought to make it "better". I found that to be a bit difficult b/c there's a good deal of "use this precedence for one param, different precedence for different param". 

I see that currently the presence of `subscription_id` info is used as a flag to use certain auth types, e.g. if `AZURE_SUBSCRIPTION_ID` is in the environment then that helps `auth_source == 'auto'` choose environment auth, even if the rest aren't necessarily defined.

So I went with making very baby steps to try and understand how everything fit together in this module and make safe changes. 

You can see what I did so far in #195. Overall I would like to make more of the auth params use precedence of "explicitly specified -> environment ->  default"  or "explict -> environment -> profile" and things like "subscription_id" apply globally with precedence "specified -> environment -> underlying auth default".

**Comment by Fred-sun** [comment]
@UnwashedMeme Recently, I have delayed my reply to you because I am busy with other things. Please help rebase #195, I will test and advance the PR merger. Thank you!

**Comment by Fred-sun** [comment]
fixed by #195

**Comment by 36roentgen** [comment]
I'm unclear about the scope of this issue as I have a similar one so not wanting to create another issue if this is still coming down the pipe.

Does this fix my problem of azure_rm_virtualmachine not using the subscription_id in the task, but the one from service principle credentials file? It ends up looking for the resource group in the subscription from the credentials file, but the service principal is permissioned across subscriptions.  In other words, can a service principal created in one subscription now be used to manage resources in a different subscription with Ansible?
